src/Canvas.cc

```diff
@@ -557,69 +519,51 @@ streamPNG(void *c, const uint8_t *data, unsigned len) {
  * StreamPngSync(this, options: {palette?: Uint8ClampedArray, backgroundIndex?: uint32, compressionLevel: uint32, filters: uint32})
  */
 
-NAN_METHOD(Canvas::StreamPNGSync) {
-  if (!info[0]->IsFunction())
-    return Nan::ThrowTypeError("callback function required");
-
-  Canvas *canvas = Nan::ObjectWrap::Unwrap<Canvas>(info.This());
+void
+Canvas::StreamPNGSync(const Napi::CallbackInfo& info) {
+  if (!info[0].IsFunction()) {
+    Napi::TypeError::New(env, "callback function required").ThrowAsJavaScriptException();
+    return;
+  }
 
-  PngClosure closure(canvas);
+  PngClosure closure(this);
   parsePNGArgs(info[1], closure);
 
-  closure.cb.Reset(Local<Function>::Cast(info[0]));
-
-  Nan::TryCatch try_catch;
+  closure.cb = Napi::Persistent(info[0].As<Napi::Function>());
 
-  cairo_status_t status = canvas_write_to_png_stream(canvas->surface(), streamPNG, &closure);
+  cairo_status_t status = canvas_write_to_png_stream(surface(), streamPNG, &closure);
 
-  if (try_catch.HasCaught()) {
-    try_catch.ReThrow();
-    return;
-  } else if (status) {
-    Local<Value> argv[1] = { Canvas::Error(status) };
-    Nan::Call(closure.cb, Nan::GetCurrentContext()->Global(), sizeof argv / sizeof *argv, argv);
-  } else {
-    Local<Value> argv[3] = {
-        Nan::Null()
-      , Nan::Null()
-      , Nan::New<Uint32>(0) };
-    Nan::Call(closure.cb, Nan::GetCurrentContext()->Global(), sizeof argv / sizeof *argv, argv);
+  if (!env.IsExceptionPending()) {
+    if (status) {
+      closure.cb.Call(env.Global(), { CairoError(status).Value() });
+    } else {
+      closure.cb.Call(env.Global(), { env.Null(), env.Null(), Napi::Number::New(env, 0) });
+    }
   }
-  return;
 }
 
 
 struct PdfStreamInfo {
-  Local<Function> fn;
+  Napi::Function fn;
   uint32_t len;
   uint8_t* data;
 };
 
-
-/*
- * Canvas::StreamPDF FreeCallback
- */
-
-void stream_pdf_free(char *, void *) {}
-
 /*
  * Canvas::StreamPDF callback.
  */
 
 static cairo_status_t
 streamPDF(void *c, const uint8_t *data, unsigned len) {
-  Nan::HandleScope scope;
-  Nan::AsyncResource async("canvas:StreamPDF");
   PdfStreamInfo* streaminfo = static_cast<PdfStreamInfo*>(c);
+  Napi::Env env = streaminfo->fn.Env();
+  Napi::HandleScope scope(env);
+  Napi::AsyncContext async(env, "canvas:StreamPDF");
   // TODO this is technically wrong, we're returning a pointer to the data in a
   // vector in a class with automatic storage duration. If the canvas goes out
   // of scope while we're in the handler, a use-after-free could happen.
-  Local<Object> buf = Nan::NewBuffer(const_cast<char *>(reinterpret_cast<const char *>(data)), len, stream_pdf_free, 0).ToLocalChecked();
-  Local<Value> argv[3] = {
-      Nan::Null()
-    , buf
-    , Nan::New<Number>(len) };
-  async.runInAsyncScope(Nan::GetCurrentContext()->Global(), streaminfo->fn, sizeof argv / sizeof *argv, argv);
+  Napi::Object buf = Napi::Buffer<uint8_t>::New(env, (uint8_t *)(data), len);
```

Comment:
Here and everywhere there's `Napi::Object xx = Napi::Buffer...`, including the return type of `Canvas::ToBuffer`, should be
```suggestion
  auto buf = Napi::Buffer<uint8_t>::New(env, (uint8_t *)(data), len);
```
or `Napi::Buffer<uint8_t>` to avoid casting to the superclass.

(Avoids an MSVC warning https://learn.microsoft.com/en-us/cpp/code-quality/C26437.)

Reply:
`Napi::Value` should work too if I understand correctly.

I don't know our stance on `auto` but I prefer seeing the types most of the time. Maybe I should set up type checking in my editor.

Reply:
`Napi::Value` is still an ancestor class. Only `auto` or `Napi::Bufer<uint8_t>` avoid the implicit cast.

Reply:
The link above is about [slicing](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-slice), which is slightly different than an implicit cast, but I see what you're saying. But wouldn't there be a slice happening every time you return a more derived type in a node-addon-api callback, which must have a return type of `Napi::Value`? It is designed to be used this way, so I'm not sure if it's really an issue. I don't see that MSVC warning on CI either.

Reply:
Implicit cast *to a super class*, right. It's not actually an issue because the only members in the derived class are retrieved from V8 when needed. Fine to ignore.